# Rain Prediction using Different Classification Model

Created By: AKASH SHAW

Created On: July 8,2020

### Contents

**Importing the Libraries**

**Importing the Dataset**

**Data Preprocessing**

**Spliting the Dataset into Training set and Test set**

**Training the Random Forest Classification Model**

**Training the Logistic Regression Model**

**Training the Decision Tree Classification Model**

**Conclusion**

## Importing the Libraries

In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns


## Importing the Dataset

In [2]:
dataset = pd.read_csv('weatherAUS.csv')

In [3]:
dataset.head()


Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  Pressure9am  \
0           W           44.0          W  ...        22.0       1007.7   
1         WNW           44.0        NNW  ...        25.0       1010.6   
2         WSW           46.0          W  ...        30.0       1007.6   
3          NE           24.0         SE  ...        16.0       1017.6   
4           W           41.0        ENE  ...        33.0       1010.8   

   Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RISK_MM  \
0       1007.1       8.0       NaN     16.9     21.8         No      0.0   
1       1007.8       NaN       NaN     17.2     24.3         No      0.0   
2       1008.7       NaN       2.0     21.0     23.2         No      0.0   
3       1012.8       NaN       NaN     18.1     26.5         No      1.0   
4       1006.0       7.0       8.0     17.8     29.7         No      0.2   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 24 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 142193 entries, 0 to 142192
Data columns (total 24 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           142193 non-null  object 
 1   Location       142193 non-null  object 
 2   MinTemp        141556 non-null  float64
 3   MaxTemp        141871 non-null  float64
 4   Rainfall       140787 non-null  float64
 5   Evaporation    81350 non-null   float64
 6   Sunshine       74377 non-null   float64
 7   WindGustDir    132863 non-null  object 
 8   WindGustSpeed  132923 non-null  float64
 9   WindDir9am     132180 non-null  object 
 10  WindDir3pm     138415 non-null  object 
 11  WindSpeed9am   140845 non-null  float64
 12  WindSpeed3pm   139563 non-null  float64
 13  Humidity9am    140419 non-null  float64
 14  Humidity3pm    138583 non-null  float64
 15  Pressure9am    128179 non-null  float64
 16  Pressure3pm    128212 non-null  float64
 17  Cloud9am       88536 non-null

## Data Preprocessing

In [5]:
dataset.isnull().sum()

Date                 0
Location             0
MinTemp            637
MaxTemp            322
Rainfall          1406
Evaporation      60843
Sunshine         67816
WindGustDir       9330
WindGustSpeed     9270
WindDir9am       10013
WindDir3pm        3778
WindSpeed9am      1348
WindSpeed3pm      2630
Humidity9am       1774
Humidity3pm       3610
Pressure9am      14014
Pressure3pm      13981
Cloud9am         53657
Cloud3pm         57094
Temp9am            904
Temp3pm           2726
RainToday         1406
RISK_MM              0
RainTomorrow         0
dtype: int64

In [6]:
dataset.shape

(142193, 24)

In [7]:
#drop the columns that will have no affect on the prediction of the model
dataset.drop(labels = ['Date','Location','Evaporation','Sunshine','Cloud3pm','Cloud9am','RISK_MM'],axis = 1,inplace = True)

In [8]:
dataset.head()

MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed WindDir9am  \
0     13.4     22.9       0.6           W           44.0          W   
1      7.4     25.1       0.0         WNW           44.0        NNW   
2     12.9     25.7       0.0         WSW           46.0          W   
3      9.2     28.0       0.0          NE           24.0         SE   
4     17.5     32.3       1.0           W           41.0        ENE   

  WindDir3pm  WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  \
0        WNW          20.0          24.0         71.0         22.0   
1        WSW           4.0          22.0         44.0         25.0   
2        WSW          19.0          26.0         38.0         30.0   
3          E          11.0           9.0         45.0         16.0   
4         NW           7.0          20.0         82.0         33.0   

   Pressure9am  Pressure3pm  Temp9am  Temp3pm RainToday RainTomorrow  
0       1007.7       1007.1     16.9     21.8        No           No  
1       1010.6       1007.8     17.2     24.3        No           No  
2       1007.6       1008.7     21.0     23.2        No           No  
3       1017.6       1012.8     18.1     26.5        No           No  
4       1010.8       1006.0     17.8     29.7        No           No

In [9]:
#Converting Predictions to Binary
## we can change that day and next days'predictions (yes and no) to 1 and 0:
dataset['RainToday'].replace({'No':0,'Yes':1},inplace = True)
dataset['RainTomorrow'].replace({'No':0,'Yes':1},inplace = True)
dataset.shape

(142193, 17)

In [10]:
dataset.head()

MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed WindDir9am  \
0     13.4     22.9       0.6           W           44.0          W   
1      7.4     25.1       0.0         WNW           44.0        NNW   
2     12.9     25.7       0.0         WSW           46.0          W   
3      9.2     28.0       0.0          NE           24.0         SE   
4     17.5     32.3       1.0           W           41.0        ENE   

  WindDir3pm  WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  \
0        WNW          20.0          24.0         71.0         22.0   
1        WSW           4.0          22.0         44.0         25.0   
2        WSW          19.0          26.0         38.0         30.0   
3          E          11.0           9.0         45.0         16.0   
4         NW           7.0          20.0         82.0         33.0   

   Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  
0       1007.7       1007.1     16.9     21.8        0.0             0  
1       1010.6       1007.8     17.2     24.3        0.0             0  
2       1007.6       1008.7     21.0     23.2        0.0             0  
3       1017.6       1012.8     18.1     26.5        0.0             0  
4       1010.8       1006.0     17.8     29.7        0.0             0

In [11]:
dataset.dropna(inplace = True)

In [12]:
dataset.shape

(112925, 17)

In [13]:
categorical = ['WindGustDir','WindDir9am','WindDir3pm']

In [14]:
# Using get_dummies for one hot encoding
dataset = pd.get_dummies(dataset,columns = categorical,drop_first=True)

In [15]:
dataset.head()

MinTemp  MaxTemp  Rainfall  WindGustSpeed  WindSpeed9am  WindSpeed3pm  \
0     13.4     22.9       0.6           44.0          20.0          24.0   
1      7.4     25.1       0.0           44.0           4.0          22.0   
2     12.9     25.7       0.0           46.0          19.0          26.0   
3      9.2     28.0       0.0           24.0          11.0           9.0   
4     17.5     32.3       1.0           41.0           7.0          20.0   

   Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  ...  WindDir3pm_NNW  \
0         71.0         22.0       1007.7       1007.1  ...               0   
1         44.0         25.0       1010.6       1007.8  ...               0   
2         38.0         30.0       1007.6       1008.7  ...               0   
3         45.0         16.0       1017.6       1012.8  ...               0   
4         82.0         33.0       1010.8       1006.0  ...               0   

   WindDir3pm_NW  WindDir3pm_S  WindDir3pm_SE  WindDir3pm_SSE  WindDir3pm_SSW  \
0              0             0              0               0               0   
1              0             0              0               0               0   
2              0             0              0               0               0   
3              0             0              0               0               0   
4              1             0              0               0               0   

   WindDir3pm_SW  WindDir3pm_W  WindDir3pm_WNW  WindDir3pm_WSW  
0              0             0               1               0  
1              0             0               0               1  
2              0             0               0               1  
3              0             0               0               0  
4              0             0               0               0  

[5 rows x 59 columns]

In [16]:
dataset.shape

(112925, 59)

In [17]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
#Excluding Tomorrow's Prediction from the Dataset
# X is the independent variable and y is the dependent variable
X = dataset.drop(labels = ['RainTomorrow'],axis = 1)
X.shape

(112925, 58)

In [18]:
X.head()

MinTemp  MaxTemp  Rainfall  WindGustSpeed  WindSpeed9am  WindSpeed3pm  \
0     13.4     22.9       0.6           44.0          20.0          24.0   
1      7.4     25.1       0.0           44.0           4.0          22.0   
2     12.9     25.7       0.0           46.0          19.0          26.0   
3      9.2     28.0       0.0           24.0          11.0           9.0   
4     17.5     32.3       1.0           41.0           7.0          20.0   

   Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  ...  WindDir3pm_NNW  \
0         71.0         22.0       1007.7       1007.1  ...               0   
1         44.0         25.0       1010.6       1007.8  ...               0   
2         38.0         30.0       1007.6       1008.7  ...               0   
3         45.0         16.0       1017.6       1012.8  ...               0   
4         82.0         33.0       1010.8       1006.0  ...               0   

   WindDir3pm_NW  WindDir3pm_S  WindDir3pm_SE  WindDir3pm_SSE  WindDir3pm_SSW  \
0              0             0              0               0               0   
1              0             0              0               0               0   
2              0             0              0               0               0   
3              0             0              0               0               0   
4              1             0              0               0               0   

   WindDir3pm_SW  WindDir3pm_W  WindDir3pm_WNW  WindDir3pm_WSW  
0              0             0               1               0  
1              0             0               0               1  
2              0             0               0               1  
3              0             0               0               0  
4              0             0               0               0  

[5 rows x 58 columns]

In [19]:
y = dataset['RainTomorrow']
X = sc.fit_transform(X)
print(X)

[[ 0.11756741 -0.10822071 -0.20666127 ... -0.27515781  3.90999267
  -0.27186642]
 [-0.84180219  0.20684494 -0.27640495 ... -0.27515781 -0.25575496
   3.67827705]
 [ 0.03761995  0.29277194 -0.27640495 ... -0.27515781 -0.25575496
   3.67827705]
 ...
 [-1.44940294  0.23548728 -0.27640495 ... -0.27515781 -0.25575496
  -0.27186642]
 [-1.16159206  0.46462594 -0.27640495 ... -0.27515781  3.90999267
  -0.27186642]
 [-0.77784422  0.4789471  -0.27640495 ... -0.27515781 -0.25575496
  -0.27186642]]


## Splitting the dataset into the Training set and Test set

In [20]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state = 0)

## Training the Random Forest Classification Model

In [21]:
from sklearn.ensemble import RandomForestClassifier
rc = RandomForestClassifier(n_estimators = 200, random_state = 0)
rc.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [22]:
#Predicting the test set results
y_pred = rc.predict(X_test)
print(y_pred)


[0 0 0 ... 0 0 0]


In [23]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score,f1_score
#Making the confusion matrix
print(confusion_matrix(y_test,y_pred))


[[16828   822]
 [ 2433  2502]]


In [24]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.87      0.95      0.91     17650
           1       0.75      0.51      0.61      4935

    accuracy                           0.86     22585
   macro avg       0.81      0.73      0.76     22585
weighted avg       0.85      0.86      0.84     22585



In [25]:
print('F1_score  ---->', f1_score(y_test, y_pred, average='weighted'))
print('Accuracy  ---->',accuracy_score(y_test,y_pred))

F1_score  ----> 0.8449667057572309
Accuracy  ----> 0.8558777949966793


## Training the Logistic Regression model

In [26]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [27]:
#Predicting the test set results
y_pred = classifier.predict(X_test)
print(y_pred)

[0 0 0 ... 0 0 0]


In [31]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score,f1_score
#Making the confusion matrix
print(confusion_matrix(y_test,y_pred))

[[16686   964]
 [ 2440  2495]]


In [32]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.87      0.95      0.91     17650
           1       0.72      0.51      0.59      4935

    accuracy                           0.85     22585
   macro avg       0.80      0.73      0.75     22585
weighted avg       0.84      0.85      0.84     22585



In [33]:
print('F1_score  ---->', f1_score(y_test, y_pred, average='weighted'))
print('Accuracy ---->',accuracy_score(y_test,y_pred))

F1_score  ----> 0.83905379961743
Accuracy ----> 0.8492804959043613


## Training the Decision Tree Classification Model 

In [34]:
from sklearn.tree import DecisionTreeClassifier
classifier_d = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier_d.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [35]:
#Predicting the test set results
y_pred = classifier_d.predict(X_test)
print(y_pred)

[0 1 0 ... 0 0 1]


In [36]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score,f1_score
#Making the confusion matrix
print(confusion_matrix(y_test,y_pred))


[[15235  2415]
 [ 2273  2662]]


In [37]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.87      0.86      0.87     17650
           1       0.52      0.54      0.53      4935

    accuracy                           0.79     22585
   macro avg       0.70      0.70      0.70     22585
weighted avg       0.79      0.79      0.79     22585



In [38]:
print('F1_score  ---->', f1_score(y_test, y_pred, average='weighted'))
print('Accuracy ---->',accuracy_score(y_test,y_pred))

F1_score  ----> 0.7934814124828274
Accuracy ----> 0.7924286030551251


#### As we observe from the above different models that Random forest Classification Model gives the best Accuracy score of ~86%(approx) and F1_Score of ~84%.
#### Logistic Regression has a Accuracy Score of ~85%(approx) and F1_Score of ~84% which is just slightly less than Random Forest model.
#### And  both these models work better than Decision tree Classification model that gives Accuracy score and F1_Score of ~79% respectively. 

### Hence, we can say that Random Forest Classification model will give the best results for this dataset.